envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt

```diff
@@ -31,40 +32,75 @@ class SnapshotUpdater(
-                Mono.justOrEmpty(listOf())
+
+            versions.retainGroups(cache.groups())
+            groups.forEach { group ->
+                if (group.ads) {
+                    updateSnapshotForGroup(group, lastAdsSnapshot)
+                } else {
+                    updateSnapshotForGroup(group, lastXdsSnapshot)
+                }
             }
+        }
     }
 
-    private fun updateSnapshots(states: List<LocalityAwareServicesState>) {
-        val snapshot = snapshotFactory.newSnapshot(states, ads = false)
-        val adsSnapshot = snapshotFactory.newSnapshot(states, ads = true)
+    fun groups(): Flux<UpdateResult> {
+        // see GroupChangeWatcher
+        return onGroupAdded
+                .publishOn(scheduler)
+                .map { groups ->
+                    UpdateResult(action = Action.SELECTED_GROUPS, groups = groups)
+                }
+                .onErrorResume { e ->
+                    meterRegistry.counter("snapshot-updater.groups.updates.errors").increment()
+                    logger.error("Unable to process new group", e)
+                    Mono.justOrEmpty(UpdateResult(action = Action.ERROR))
+                }
+    }
 
-        cache.groups().forEach { group -> updateSnapshotForGroup(group, if (group.ads) adsSnapshot else snapshot) }
+    fun services(changes: Flux<List<LocalityAwareServicesState>>): Flux<UpdateResult> {
+        return changes
+                .sample(properties.stateSampleDuration)
+                .publishOn(scheduler)
+                .map { states ->
+                    updateSnapshots(states)
+                    UpdateResult(action = Action.ALL_GROUPS)
+                }
+                .onErrorResume { e ->
+                    meterRegistry.counter("snapshot-updater.services.updates.errors").increment()
+                    logger.error("Unable to process service changes", e)
+                    Mono.justOrEmpty(UpdateResult(action = Action.ALL_GROUPS))
+                }
     }
 
-    private fun updateSnapshotForGroup(group: Group, snapshot: Snapshot) {
+    private fun updateSnapshots(states: List<LocalityAwareServicesState>) {
+        lastXdsSnapshot = snapshotFactory.newSnapshot(states, ads = false)
+        lastAdsSnapshot = snapshotFactory.newSnapshot(states, ads = true)
+    }
+
+    fun updateSnapshotForGroup(group: Group, globalSnapshot: Snapshot) {
         try {
-            val groupSnapshot = snapshotFactory.getSnapshotForGroup(group, snapshot)
+            val groupSnapshot = snapshotFactory.getSnapshotForGroup(group, globalSnapshot)
             cache.setSnapshot(group, groupSnapshot)
         } catch (e: Throwable) {
             meterRegistry.counter("snapshot-updater.services.${group.serviceName}.updates.errors").increment()
             logger.error("Unable to create snapshot for group ${group.serviceName}", e)
         }
     }
 }
+
+enum class Action {
+    SELECTED_GROUPS, ALL_GROUPS, ERROR
```

Comment:
This is cryptic. `Action` contains nouns. I think it should contain verbs. That would clarify the meaning.